src/page-container/usePageManager.jsx

```diff
@@ -0,0 +1,200 @@
+import React from 'react';
+
+import Logger from '../utils/logger';
+import {
+  getPersistentScrollMap,
+  applyScrollData,
+} from '../utils/scroll-persistence/scroll-persistence';
+
+import PageManagerContext from './PageManagerContext';
+
+/**
+ * Hook used to manage Page registrations within a PageContainer.
+ * @param {Object} rootContainerRef A React ref object with a current value pointing
+ * to a valid HTMLElement
+ * @returns An object containing data related to currently active/visible Page
+ * and a unique PageManager component definition
+ */
+const usePageManager = (rootContainerRef) => {
+  const pageRegisterRef = React.useRef({});
+  const lastActivePageArrayRef = React.useRef([]);
+  const [activePageArray, setActivePageArray] = React.useState([]);
+
+  const pageManagerContextValue = React.useMemo(() => {
+    const registerPage = ({
+      portalElement,
+      pageId,
+      parentPageId,
+      label,
+      metaData,
+    }) => {
+      pageRegisterRef.current[pageId] = {
+        pageId,
+        portalElement,
+        parentPageId,
+        label,
+        metaData,
+        overflowData: pageRegisterRef.current[pageId]
+          ? pageRegisterRef.current[pageId].overflowData : undefined,
+      };
+
+      setActivePageArray((activePages) => {
+        // If the specified Page is already present in the render order,
+        // it can stay where it is. The state is updated to ensure that any changed
+        // page values (label/metaData) are communicated to the other effects.
+        if (activePages.indexOf(pageId) >= 0) {
+          return [...activePages];
+        }
+
+        // If there are currently no registered Pages, we initialize the state
+        // with the single new entry.
+        if (!activePages.length) {
+          return [pageId];
+        }
+
+        // If no ancestor Page is specified, but there are known active Pages,
+        // this new Page must be a duplicate at the root of the PageContainer.
+        // This new Page is not queued for rendering and is ignored.
+        if (!parentPageId) {
+          // We need to perform a separate check to see if the new Page should
+          // be inserted into the existing ordering, likely due to a series of
+          // Pages mounting at the same time.
+          const existingChild = pageRegisterRef.current[activePages[0]];
+          if (existingChild && existingChild.parentPageId === pageId) {
+            const newPages = [...activePages];
```

Comment:
nitpick - instead of splicing, we could just initialize the array with the pageId as the first element
```suggestion
            const newPages = [pageId, ...activePages];
            return newPages;
```